### PR TITLE
Serialize and deserialize docker image repo

### PR DESCRIPTION
**Pull request by hmusum**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java [hunk 1]
```diff
@@ -187,8 +187,7 @@ private void toSlime(Allocation allocation, Cursor object) {
         object.setLong(currentRestartGenerationKey, allocation.restartGeneration().current());
         object.setBool(removableKey, allocation.isRemovable());
         object.setString(wantedVespaVersionKey, allocation.membership().cluster().vespaVersion().toString());
-        // TODO serialize dockerImageRepo
-        //object.setString(wantedDockerImageRepoKey, allocation.membership().cluster().dockerImageRepo().orElse(""));
+        object.setString(wantedDockerImageRepoKey, allocation.membership().cluster().dockerImageRepo().orElse(""));
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java]: I realize the deserializer handles empty string, but consider just to serializing dockerImageRepo if it's not set.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java [hunk 2]
```diff
@@ -157,8 +157,11 @@ private void toSlime(Node node, boolean allFields, Cursor object) {
             toSlime(allocation.membership(), object.setObject("membership"));
             object.setLong("restartGeneration", allocation.restartGeneration().wanted());
             object.setLong("currentRestartGeneration", allocation.restartGeneration().current());
-            object.setString("wantedDockerImage", nodeRepository.dockerImage(node).withTag(allocation.membership().cluster().vespaVersion()).asString());
-            object.setString("wantedVespaVersion", allocation.membership().cluster().vespaVersion().toFullString());
+            String wantedVespaVersion = allocation.membership().cluster().vespaVersion().toFullString();
+            Optional<String> dockerImageRepo = allocation.membership().cluster().dockerImageRepo();
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java]: Consider having a method on `ClusterSpec` that returns the full docker image

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java [hunk 3]
```diff
@@ -156,6 +157,33 @@ public void nodeVersionIsReturnedIfSet() {
...
+
+        // deploy
+        ApplicationId application1 = tester.makeApplicationId();
+        SystemState state1 = prepare(application1, tester, 1, 1, 1, 1, defaultResources, "1.2.3");
+        String dockerImageRepo = "docker.domain.tld/my/image";
+        prepare(application1, tester, 1, 1, 1 , 1 , false, defaultResources, "1.2.3", Optional.of(dockerImageRepo));
+        tester.activate(application1, state1.allHosts);
+
+        HostSpec host1 = state1.container0.iterator().next();
+        assertFalse(host1.version().isPresent());
```
Comment (freva) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java]: Why is this the case?

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java [hunk 4]
```diff
@@ -156,6 +157,33 @@ public void nodeVersionIsReturnedIfSet() {
...
+        Node node1 = tester.nodeRepository().getNode(host1.hostname()).get();
+        tester.nodeRepository().write(node1.with(node1.status().withDockerImage(DockerImage.fromString(dockerImageRepo))), () -> {});
```
Comment (freva) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java]: This actually shouldn't be allowed, current docker image should always include the version as well.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java (hunk 1 above)
Comment (hmusum) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java]: Yeah, fixed

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java (hunk 2 above)
Comment (hmusum) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java]: Done

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java (hunk 3 above)
Comment (hmusum) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java]: Should not have been there, removed

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java (hunk 4 above)
Comment (hmusum) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/provisioning/ProvisioningTest.java]: test code updated

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java [hunk 5]
```diff
@@ -157,8 +157,10 @@ private void toSlime(Node node, boolean allFields, Cursor object) {
             toSlime(allocation.membership(), object.setObject("membership"));
             object.setLong("restartGeneration", allocation.restartGeneration().wanted());
             object.setLong("currentRestartGeneration", allocation.restartGeneration().current());
-            object.setString("wantedDockerImage", nodeRepository.dockerImage(node).withTag(allocation.membership().cluster().vespaVersion()).asString());
-            object.setString("wantedVespaVersion", allocation.membership().cluster().vespaVersion().toFullString());
+            String wantedVespaVersion = allocation.membership().cluster().vespaVersion().toFullString();
+            object.setString("wantedDockerImage", allocation.membership().cluster().dockerImage()
+                    .orElse(nodeRepository.dockerImage(node).withTag(allocation.membership().cluster().vespaVersion()).asString()));
+            object.setString("wantedVespaVersion", wantedVespaVersion);
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodesResponse.java]:
```suggestion
            object.setString("wantedDockerImage", allocation.membership().cluster().dockerImage()
                    .orElseGet(() -> nodeRepository.dockerImage(node).withTag(allocation.membership().cluster().vespaVersion()).asString()));
            object.setString("wantedVespaVersion", allocation.membership().cluster().vespaVersion().toFullString());
```
(Changing back to `orElseGet()` and setting `wantedVespaVersion` directly)

**Review by freva — APPROVED** [review]
:+1: 

Is there any plan to default to some `dockerImageRepo` so `ClusterSpec` always has it? Or maybe just get rid of `docker-image-override` flag and use default for node type?

**Comment by hmusum** [comment]
Yes, I want to get rid of `docker-image-override` flag and use default for node type after this goes in and we have start using this instead

**Pull request merged**